File: yandex/ytcf/2018/final/c.py
n = int(input())
m = int(input())
d = [[list(map(int, input().split())) for j in range(m)] for i in range(n - 1)]
source, target = map(int, input().split())
D = int(input())

cache = {}


def f(i, j, dd):
    if i == 0:
        return j == source and dd == 0

    if dd > max_dist[i][j]:
        return False

    key = (i, j, dd)
    if key in cache:
        return cache[key]

    for k in range(m):
        if dd >= d[i - 1][k][j] and f(i - 1, k, dd - d[i - 1][k][j]):
            cache[key] = True
            return True

    cache[key] = False
    return False


max_dist = [[0 for j in range(m)] for i in range(n)]
for i in range(1, n):
    for j in range(m):
        for k in range(m):
            max_dist[i][k] = max(max_dist[i][k],
                                 max_dist[i - 1][j] + d[i - 1][j][k])

print('YES' if f(n - 1, target, D) else 'NO')

File: yandex/ytcf/2018/final/c.cpp
#include <iostream>
#include <unordered_map>
#include <algorithm>

using namespace std;

typedef unsigned long long ull;

struct Key {
    int i, j;
    ull d;

    bool operator==(const Key &other) const {
        return i == other.i && j == other.j && d == other.d;
    }
};

struct KeyHasher {
  size_t operator () (const Key &key) const {
    return hash<int>()(key.i) ^ hash<int>()(key.j) ^ hash<ull>()(key.d);
  }
};

int n, m, source, target;
ull d[22][5][5], D;
ull maxDist[22][5];
unordered_map<Key, bool, KeyHasher> cache;

bool f(const int i, const int j, const ull dd) {
    if (i == 0)
        return j == source && dd == 0;

    if (dd > maxDist[i][j])
        return false;

    Key key = {.i=i, .j=j, .d=dd};

    if (cache.find(key) != cache.end())
        return cache[key];

    for (int k = 0; k < m; k ++)
        if (dd >= d[i - 1][k][j] && f(i - 1, k, dd - d[i - 1][k][j])) {
            cache[key] = true;
            return true;
        }

    cache[key] = false;
    return false;
}


int main () {
    cin >> n >> m;

    for (int i = 0; i < n - 1; i ++)
        for (int j = 0; j < m; j ++)
            for (int k = 0; k < m; k ++)
                cin >> d[i][j][k];

    cin >> source >> target;
    cin >> D;

    for (int i = 1; i < n; i ++)
        for (int j = 0; j < m; j ++)
            for (int k = 0; k < m; k ++)
                maxDist[i][k] = max(maxDist[i][k], maxDist[i - 1][j] + d[i - 1][j][k]);

    cout << (f(n - 1, target, D) ? "YES" : "NO");
    return 0;
}